File: app/src/main/java/com/nfcalarmclock/NacDayButton.java
package com.nfcalarmclock;

import android.content.Context;
import android.content.res.Resources;
import android.content.res.TypedArray;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.LinearLayout.LayoutParams;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.button.MaterialButtonToggleGroup;

/**
 * A button that consists of an image to the left, and text to the right
 * of it.
 */
@SuppressWarnings("UnnecessaryInterfaceModifier")
public class NacDayButton
	extends LinearLayout
	implements View.OnClickListener,
		MaterialButtonToggleGroup.OnButtonCheckedListener
{

	/**
	 * Listener for day change events.
	 */
	public interface OnDayChangedListener
	{
		public void onDayChanged(NacDayButton button);
	}

	/**
	 * Attributes for button and text.
	 */
	public static class Attributes
	{
		public final int width;
		public final int height;
		public final String text;

		/**
		 * Initialize the attributes.
		 */
		public Attributes(Context context, AttributeSet attrs)
		{
			Resources.Theme theme = context.getTheme();
			TypedArray ta = theme.obtainStyledAttributes(attrs,
				R.styleable.NacDayButton, 0, R.style.NacDayButton);
			TypedArray androidTa = theme.obtainStyledAttributes(attrs,
				new int[] { android.R.attr.text }, 0, 0);

			try
			{
				this.width = (int) ta.getDimension(R.styleable.NacDayButton_nacWidth,
					ViewGroup.LayoutParams.WRAP_CONTENT);
				this.height = (int) ta.getDimension(R.styleable.NacDayButton_nacHeight,
					ViewGroup.LayoutParams.WRAP_CONTENT);
				this.text = androidTa.getString(0);
			}
			finally
			{
				ta.recycle();
			}
		}

	}

	/**
	 * Day button parent.
	 */
	private MaterialButtonToggleGroup mButtonToggleGroup;

	/**
	 * Day button.
	 */
	private MaterialButton mButton;

	/**
	 * View attributes.
	 */
	private Attributes mAttributes;

	/**
	 * Day changed listener.
	 */
	private NacDayButton.OnDayChangedListener mOnDayChangedListener;

	/**
	 */
	public NacDayButton(Context context)
	{
		super(context, null);
		init(null);
	}

	/**
	 */
	public NacDayButton(Context context, AttributeSet attrs)
	{
		super(context, attrs, 0);
		init(attrs);
	}

	/**
	 */
	public NacDayButton(Context context, AttributeSet attrs,
		int defStyleAttr)
	{
		super(context, attrs, defStyleAttr);
		init(attrs);
	}

	/**
	 * Disable and animate the view.
	 */
	public void disable()
	{
		this.getButton().setChecked(false);
	}

	/**
	 * Enable and animate the view.
	 */
	public void enable()
	{
		this.getButton().setChecked(true);
	}

	/**
	 * @return The button.
	 */
	public MaterialButton getButton()
	{
		return this.mButton;
	}

	/**
	 * @return The button toggle group.
	 */
	public MaterialButtonToggleGroup getButtonToggleGroup()
	{
		return this.mButtonToggleGroup;
	}

	/**
	 * @return The day attributes object.
	 */
	private Attributes getDayAttributes()
	{
		return this.mAttributes;
	}

	/**
	 * @return The day button on click listener.
	 */
	public NacDayButton.OnDayChangedListener getOnDayChangedListener()
	{
		return this.mOnDayChangedListener;
	}

	/**
	 * Initialize the view.
	 */
	public void init(AttributeSet attrs)
	{
		if (attrs == null)
		{
			return;
		}

		setOrientation(LinearLayout.HORIZONTAL);
		this.setupStyle();

		Context context = getContext();
		this.mAttributes = new Attributes(context, attrs);
		this.mOnDayChangedListener = null;

		this.getButtonToggleGroup().addOnButtonCheckedListener(this);
		setOnClickListener(this);
	}

	/**
	 */
	@Override
	public void onButtonChecked(MaterialButtonToggleGroup group, int checkedId,
		boolean isChecked)
	{
		NacUtility.printf("onButtonChecked!");
		NacDayButton.OnDayChangedListener listener = this.getOnDayChangedListener();
		if (listener != null)
		{
			listener.onDayChanged(this);
		}
	}

	/**
	 */
	@Override
	public void onClick(View view)
	{
		NacUtility.printf("onClick!");
		MaterialButton button = this.getButton();
		boolean checked = button.isChecked();

		button.setChecked(!checked);
	}

	/**
	 * Finish setting up the View.
	 */
	@Override
	protected void onFinishInflate()
	{
		super.onFinishInflate();
		this.setViewAttributes();
	}

	/**
	 * Set the listener for when the day is changed.
	 */
	public void setOnDayChangedListener(NacDayButton.OnDayChangedListener listener)
	{
		this.mOnDayChangedListener = listener;
	}

	/**
	 * Set the width and height of the button.
	 */
	public void setSize(int width, int height)
	{
		MaterialButton button = this.getButton();
		LayoutParams params = (LayoutParams) button.getLayoutParams();
		params.width = width;
		params.height = height;

		button.setLayoutParams(params);
	}

	/**
	 * Set the day button style.
	 */
	public void setStyle(int style)
	{
		Context context = getContext();
		LayoutInflater inflater = LayoutInflater.from(context);
		int layout = (style == 1)
			? R.layout.nac_day_button_filled
			: R.layout.nac_day_button_outlined;

		removeAllViews();
		inflater.inflate(layout, this, true);

		this.mButtonToggleGroup = findViewById(R.id.nac_day_button_group);
		this.mButton = findViewById(R.id.nac_day_button);
	}

	/**
	 * Set the text in the button.
	 *
	 * @param  text  The text to enter in the button.
	 */
	public void setText(String text)
	{
		this.getButton().setText(text);
	}

	/**
	 * Setup the style of the day button.
	 */
	protected void setupStyle()
	{
		Context context = getContext();
		NacSharedPreferences shared = new NacSharedPreferences(context);
		int style = shared.getDayButtonStyle();

		this.setStyle(style);
	}

	/**
	 * Set view attributes.
	 */
	public void setViewAttributes()
	{
		Attributes attrs = this.getDayAttributes();

		this.setSize(attrs.width, attrs.height);
		this.setText(attrs.text);
	}

}

File: app/src/main/java/com/nfcalarmclock/view/dayofweek/NacDayButton.kt
package com.nfcalarmclock.view.dayofweek

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import com.google.android.material.button.MaterialButton
import com.google.android.material.button.MaterialButtonToggleGroup
import com.google.android.material.button.MaterialButtonToggleGroup.OnButtonCheckedListener
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedPreferences

/**
 * A button that consists of an image to the left, and text to the right
 * of it.
 */
class NacDayButton
	: LinearLayout,
	View.OnClickListener,
	OnButtonCheckedListener
{

	/**
	 * Listener for day change events.
	 */
	interface OnDayChangedListener
	{
		fun onDayChanged(button: NacDayButton)
	}

	/**
	 * Attributes for button and text.
	 */
	class Attributes(context: Context, attrs: AttributeSet?)
	{

		/**
		 * Width.
		 */
		var width = 0

		/**
		 * Height.
		 */
		var height = 0

		/**
		 * Text.
		 */
		var text: String? = null

		/**
		 * Initialize the attributes.
		 */
		init
		{
			// Create a typed array to search for custom attributes
			val ta = context.theme.obtainStyledAttributes(attrs,
				R.styleable.NacDayButton, 0, R.style.NacDayButton)

			// Create a typed array searching for default android attributes
			val androidTa = context.theme.obtainStyledAttributes(attrs, intArrayOf(android.R.attr.text), 0, 0)

			// Get attributes from the typed array
			try
			{
				width = ta.getDimension(R.styleable.NacDayButton_nacWidth,
					ViewGroup.LayoutParams.WRAP_CONTENT.toFloat()).toInt()
				height = ta.getDimension(R.styleable.NacDayButton_nacHeight,
					ViewGroup.LayoutParams.WRAP_CONTENT.toFloat()).toInt()
				text = androidTa.getString(0)
			}
			// Recycle the typed array
			finally
			{
				ta.recycle()
			}
		}

	}

	/**
	 * Day button parent.
	 */
	private var buttonToggleGroup: MaterialButtonToggleGroup? = null

	/**
	 * Day button.
	 */
	var button: MaterialButton? = null
		private set

	/**
	 * View attributes.
	 */
	private var dayAttributes: Attributes? = null

	/**
	 * Day changed listener.
	 */
	var onDayChangedListener: OnDayChangedListener? = null

	/**
	 * Constructor.
	 */
	constructor(context: Context) : super(context, null)
	{
		init(null)
	}

	/**
	 * Constructor.
	 */
	constructor(context: Context, attrs: AttributeSet?) : super(context, attrs, 0)
	{
		init(attrs)
	}

	/**
	 * Constructor.
	 */
	constructor(context: Context, attrs: AttributeSet?,
		defStyleAttr: Int) : super(context, attrs, defStyleAttr)
	{
		init(attrs)
	}

	/**
	 * Disable and animate the view.
	 */
	fun disable()
	{
		// Toggle the button if it is enabled
		if (button!!.isChecked)
		{
			toggle()
		}
	}

	/**
	 * Enable and animate the view.
	 */
	fun enable()
	{
		// Toggle the button if it is disabled
		if (!button!!.isChecked)
		{
			toggle()
		}
	}

	/**
	 * Initialize the view.
	 */
	fun init(attrs: AttributeSet?)
	{
		// Check if attributes are null
		if (attrs == null)
		{
			return
		}

		// Set the orientation
		orientation = HORIZONTAL

		// Setup style
		setupStyle()

		// Set other attributes
		dayAttributes = Attributes(context, attrs)
		onDayChangedListener = null
		buttonToggleGroup!!.addOnButtonCheckedListener(this)

		// Set an on click listener
		setOnClickListener(this)
	}

	/**
	 * Called when the button is checked.
	 */
	override fun onButtonChecked(group: MaterialButtonToggleGroup, checkedId: Int,
		isChecked: Boolean)
	{
		// Return if there are no listeners set
		val listener = onDayChangedListener ?: return

		// Remove the toggle group listener that was set when this view is
		// initialized so that this onButtonChecked() method does not potentially get
		// called recursively
		buttonToggleGroup!!.removeOnButtonCheckedListener(this)

		// Call the listener
		listener.onDayChanged(this)

		// Add the toggle group listener that was set when this view is
		// initialized back
		buttonToggleGroup!!.addOnButtonCheckedListener(this)
	}

	/**
	 * Called when the view is clicked
	 */
	override fun onClick(view: View)
	{
		// Get the current state
		val checked = button!!.isChecked

		// Toggle the state
		button!!.isChecked = !checked
	}

	/**
	 * Finish setting up the View.
	 */
	override fun onFinishInflate()
	{
		super.onFinishInflate()

		setViewAttributes()
	}

	/**
	 * Set the width and height of the button.
	 */
	fun setSize(width: Int, height: Int)
	{
		// Get the button's current layout parameters
		val params = button!!.layoutParams as LayoutParams

		// Change the parameters
		params.width = width
		params.height = height

		// Apply the parameters to the button
		button!!.layoutParams = params
	}

	/**
	 * Set the day button style.
	 */
	fun setStyle(style: Int)
	{
		// Get the layout inflater
		val inflater = LayoutInflater.from(context)

		// Determine which layout to use
		val layout = if (style == 1)
		{
			R.layout.nac_day_button_filled
		}
		else
		{
			R.layout.nac_day_button_outlined
		}

		// Remove all views
		removeAllViews()

		// Inflate the layout
		inflater.inflate(layout, this, true)

		// Set the member variables
		buttonToggleGroup = findViewById(R.id.nac_day_button_group)
		button = findViewById(R.id.nac_day_button)
	}

	/**
	 * Set the text in the button.
	 *
	 * @param  text  The text to enter in the button.
	 */
	fun setText(text: String?)
	{
		button!!.text = text
	}

	/**
	 * Setup the style of the day button.
	 */
	private fun setupStyle()
	{
		// Get the style
		val shared = NacSharedPreferences(context)
		val style = shared.dayButtonStyle

		// Set the style
		setStyle(style)
	}

	/**
	 * Set view attributes.
	 */
	private fun setViewAttributes()
	{
		// Set size
		setSize(dayAttributes!!.width, dayAttributes!!.height)

		// Set text
		setText(dayAttributes!!.text)
	}

	/**
	 * Toggle the day button.
	 */
	fun toggle()
	{
		// Get the current status
		val status = button!!.isChecked

		// Toggle the status
		button!!.isChecked = !status
	}

}